nemo/collections/tts/models/vits.py

```diff
@@ -0,0 +1,425 @@
+                self._cfg.hop_size,
+                self._cfg.preprocessor.n_window_size,
+                self._cfg.mel_fmin,
+                self._cfg.mel_fmax
+            )
+            y = torch.unsqueeze(y, 1)
+            y = modules.slice_segments(y, ids_slice * self._cfg.hop_size, self._cfg.segment_size)  # slice
+            y_d_hat_r, y_d_hat_g, _, _ = self.net_d(y, y_hat.detach())
+            loss_disc, losses_disc_r, losses_disc_g = self.disc_loss(y_d_hat_r, y_d_hat_g)
+            loss_disc_all = loss_disc
+
+        # train discriminator
+        self.optim_d.zero_grad()
+        self.manual_backward(loss_disc_all)
+        modules.clip_grad_value_(self.net_d.parameters(), None)
+        self.optim_d.step()
+
+        with autocast(enabled=True):
+            # Generator
+            y_d_hat_r, y_d_hat_g, fmap_r, fmap_g = self.net_d(y, y_hat)
+        with autocast(enabled=False):
+            loss_dur = torch.sum(l_length.float())
+            loss_mel = F.l1_loss(y_mel, y_hat_mel) * self._cfg.c_mel
+            loss_kl = self.kl_loss(z_p=z_p, logs_q=logs_q, m_p=m_p, logs_p=logs_p, z_mask=z_mask) * self._cfg.c_kl
+            loss_fm = self.feat_matching_loss(fmap_r=fmap_r, fmap_g=fmap_g)
+            loss_gen, losses_gen = self.gen_loss(disc_outputs=y_d_hat_g)
+            loss_gen_all = loss_gen + loss_fm + loss_mel + loss_dur + loss_kl
+
+        # train generator
+        self.optim_g.zero_grad()
+        self.manual_backward(loss_gen_all)
+        modules.clip_grad_value_(self.net_g.parameters(), None)
+        self.optim_d.step()
+
+        schedulers = self.lr_schedulers()
+        if schedulers is not None:
+            sch1, sch2 = schedulers
+            sch1.step()
+            sch2.step()
+
+        metrics = {
+            "loss_gen": loss_gen,
+            "loss_fm": loss_fm,
+            "loss_mel * c_mel": loss_mel,
+            "loss_dur": loss_dur,
+            "loss_kl * c_kl": loss_kl,
+            "loss_gen_all": loss_gen_all,
+            "losses_disc_r": losses_disc_r,
+            "losses_disc_g": losses_disc_g,
+            "loss_disc_all": loss_disc_all,
+        }
+        
+        # TODO: Fix logging
+        # self.log_dict(metrics, on_step=True, sync_dist=True)
+
+    def validation_step(self, batch, batch_idx):
+        # (x, x_lengths, spec, spec_lengths, y, y_lengths) = batch
+        (y, y_lengths, x, x_lengths) = batch
+
+        y_hat, attn, mask, *_ = self.net_g.infer(x, x_lengths, max_len=1000)
+        y_hat = y_hat.squeeze()
+        y_hat_lengths = mask.sum([1, 2]).long() * self._cfg.train_ds.dataset.hop_length
+
+        # spec = self.get_spec(y)
+        # spec_lengths = torch.ones(spec.shape[0]) * spec.shape[2]
+
+        # Note to modify the functions / use the ones in NeMo, we need the lengths
+        mel, mel_lengths = self.audio_to_melspec_precessor(y, y_lengths)
+        y_hat_mel, y_hat_mel_lengths = self.audio_to_melspec_precessor(y_hat, y_hat_lengths)
```

Comment:
What is the difference between `self.audio_to_melspec_precessor`, and the `modules.spec_to_mel_torch`/`modules.mel_spectrogram_torch` functions?